cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,8 +738,57 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
   return Status::Invalid("Unrecognized filesystem type in URI: ", uri_string);
 }
 
+Status CheckFileSystem(const FileSystem* filesystem,
+                       const std::string& expected_type_name, const std::string& uri) {
+  if (filesystem && filesystem->type_name() != expected_type_name) {
+    return Status::Invalid("URI ", uri, " is not compatible with filesystem of type ",
+                           filesystem->type_name());
+  }
+  return Status::OK();
+}
+
 }  // namespace
 
+Result<std::string> PathFromUriOrPath(const FileSystem* filesystem,
```

Comment:
I suppose this feels like it should be a virtual method on FileSystem to me, still. (Probably with a default implementation that returns NotImplemented.) It would avoid the ifdef tower, and would keep FileSystem as an actual extension point, avoiding a hardcoded list that we have to maintain inside of Arrow. It effectively already is implemented as a (static) method, too.

Then again, other methods above have this same pattern, so I suppose it's consistent.

Reply:
Ah, can't move the existing methods because you don't have a filesystem at that point.  I'll update this one to be a virtual method.